app/src/main/java/com/kodeco/android/countryinfo/repositories/CountryRepositoryImpl.kt

```diff
@@ -0,0 +1,32 @@
+import com.kodeco.android.countryinfo.models.Country
+import com.kodeco.android.countryinfo.models.CountryFlags
+import com.kodeco.android.countryinfo.models.CountryName
+import com.kodeco.android.countryinfo.network.CountryService
+import com.kodeco.android.countryinfo.ui.screens.countryinfo.CountryInfoState
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+class CountryRepositoryImpl(private val service: CountryService) : CountryRepository {
+
+    private var countriesList: List<Country> = emptyList()
+
+    override fun fetchCountries(uptimeCounter: Int): Flow<CountryInfoState> {
+        return flow {
+            emit(CountryInfoState.Loading(uptimeCounter))
+            delay(1_000) // Simulate delay
+            val countriesResponse = service.getAllCountries()
+            val newState = if (countriesResponse.isSuccessful) {
+                countriesList = countriesResponse.body()!!
+                CountryInfoState.Success(countriesResponse.body()!!)
+            } else {
+                CountryInfoState.Error(Throwable("Request failed: ${countriesResponse.message()}"))
+            }
+            emit(newState)
+        }
+    }
+
+    override suspend fun getCountry(uniqueId: String): Country? {
+        return countriesList.find { it.name == CountryName(uniqueId) }
```

Comment:
For a slight optimization you could change the `countriesList` to a `countriesMap: Map<Country>`.

And adjust this code:

```kotlin
countriesList = countriesResponse.body()!!
```

To:

```kotlin
countriesMap = countriesResponse.body()!!.associateBy {
    { it.name }, // Lambda for the "key"
    { it }, Lambda for the value
}
```

Then in `getCountry()` you can do:

```kotlin
return countriesMap.getOrNull(uniqueId)
```